packages/presets/client/src/fragment-masking-plugin.ts

```diff
@@ -13,10 +24,43 @@ export type FragmentType<TDocumentType extends DocumentTypeDecoration<any, any>>
   : never;`;
 
 const makeFragmentDataHelper = `
+type UnionToIntersectGroupByTypeName<U, V = U> = [V] extends [
+  { __typename?: infer TypeName }
+]
+  ? TypeName extends any
+    ? UnionToIntersection<U extends { __typename?: TypeName } ? U : never>
+    : never
+  : never;
+
+type UnionFieldToIntersection<T> = [T] extends [never] ? never
+  : [T] extends [Array<unknown>]
+  ? Array<
+      | UnionFieldToIntersection<ExcludePrimitive<T[number]>>
+      | ExtractPrimitive<T[number]>
+    >
+  : UnionToIntersectGroupByTypeName<T> extends infer V
+  ? {
+      [Key in keyof V]:
+        | UnionFieldToIntersection<ExcludePrimitive<V[Key]>>
+        | ExtractPrimitive<V[Key]>;
+    }
+  : never;
+
+type Flatten<F> = [F] extends [never] ? never
+  : F extends Array<unknown>
+  ? Array<Flatten<ExcludePrimitive<F[number]>> | ExtractPrimitive<F[number]>>
+  : {
+      [Key in keyof Omit<F, " $fragmentRefs" | " $fragmentName">]:
+        | Flatten<ExcludePrimitive<F[Key]>>
+        | ExtractPrimitive<F[Key]>;
+    } & (F extends { " $fragmentRefs"?: { [K in string]: infer FRefs }; }
+      ? (FRefs extends any ? Flatten<FRefs> : never) : {}
+    );
+export type UnmaskFragment<F> = UnionFieldToIntersection<Flatten<F>>;
```

Comment:
`UnmaskFragment` utility type consists of two utility types, `Flatten` and `UnionFieldToIntersection`.

`Flatten` refers to `' $fragmentRefs'` properties and returns a new type by making the all sub-fragment flat recursively.
it returns a union type of fragments if the fragment includes multiple fragments in the same field. So we use `UnionFieldToIntersection` to transform it to an intersection type of them.

I made [a playground](https://www.typescriptlang.org/play?ts=5.2.2#code/C4TwDgpgBACgTgSwLYOAgbtAvFAzsRAOwHMoAfKQgVyQCMI5ypaB7FgGwgENCnaFiCQsCa4QdDk2rt2TKoQAmEAGZCICgNwAoUJCgBRAB4BjdlSXxkqDBAA8AFQB8UHEdPm79gDSxEKNJiO2rrQRgRcxsCW-jYOzq4mZkoOPm5Jnj7R1oFBWjrg0ACqhAgshPYsAJLCDLgQkaWEtoXxUAAUhVAQhsAQirhQPCBQAPztANYAXFCFAJQuzugsCApQ04QQmHDz3b397VpQUFNQQsoMUJVa81iLywqHo5eP65sM2nkhMyVlFdW9cDqkQA4nAWFQwAAhED2AoAOS4SDshR8ADUXDNWgBtVEAXS6PT6CgGWIA3lAAPoUkKEREQEbTM4XWGQBFIqAAX1xjzGLIgbOguyJAyGjyOY2KjT+NUB9TQZWaBL2xKg5KpNLpDKgfIFnKenVeW0cYrWlDecBeZq2Hy+krKADEEBB2AppQCgfKmk4MVj7PihfssRsttzxVaGJbff7CYGAIJwOBcEC2eTjQgsADuhEcoae8cTybthEdztdVRlHsatjSHiyAU8QZo9DguMczgoYUTkTrsXsjboDFbxqO0yLbtqctB4KhMPhdLiSuFp0I50YqJ5qpNRyxAGkIMMhMd9yxlFA8aOfsWnS7x7KGgqaxY-Nk7Kjd-uh0xOxEos-67Y3z3EBW20I4jg5S1g3eT4Cige12C4YA9lse1sXtaNlRJKCWw3bDLXtRd9nzJMU0INNM2zDdiOTeDEOQx8IB7TAUP7ZtPw7Hou1-Kx-3tVjBzbS1SS3KB3wPXhxmPU8AHl-BQnwAHIoAAEmURNiCRYQACUVFwJSKCU1T1M04ABQUnNJhEo4KFopC+mrRJaz-WI+KAocrK-TifyYuxXI-XIwL1AAydoCIDFVyUMtSuA0vpgB05Q9K1cld2XPACCEYhcUZFcLntBKBg5TkRLGfLdMIlVRUCwLSoQuymjKxLh2qkdwwtFrplJDlZm0bowBYOARFtQgkC4XBxntYy4pQ1oixLG9y3dOUq1s5DUNyPJNoAei2qAAFoDsOo7jpO06Tq0Hb9sO7UIHwKBjDG26rrOs6Lt2l6Ps+o68kuo7tVg+xbpEQo0HYVBhk+mC9H0ABHKguHYWwAA0fAATXiR42jiNobmcewKoGJGngARlNAAmHYY2JTHsdx7UCagVGSfJ+YxgIKhoGmZQEbqKHQkMSBImRhn2YgVokY+N7nr2qAAFkGGIaAvsh-I9EKOo4FJrBHgoYSWvVApaSRLUFPVhgFNA6qVmmfAiGIS3AqNiAbYykgHeK6zN316lDc1aZTY1i2ROt9K7fdo4Yudygm2gwKOWCWCzc1nTcCodgRG1z29eqg3ICdk2k6DlqQ9tzLw8oOkXbDkTI-WGO4Hd+O+ZmDWyZcHWvZzn2879qAA-N8uS9d+2ROMMpgB-ZKPNzvpe4UgBhcefyLlrTgUKuy48iBRoQdgN7djywAACzKel99IChpFkCh5CUVQNgeFqisvtPr6gW+VDUR+wIgrORJn-O-tC6D3XqHTeLUx7CEnp1ae3dZ7G39ovKBkQV7F1AaXA+q9lz4BiomJAWoMEX2jjIOQihP4P3Ls-Yhb8P733UCaJutpW4pzThnE02dAoALnsA4O6Dh7l0gRPSIU8sFcIQX3JBQjgCoKtnw6uWDt5cF3ufcuRxj6nwIcPKQr9SF3y-qo7BE9iB4M0XbbRJCb5kLod-OO5iaFWK-gwyWqsigawAMztz-t7DU4j+5wBkUcIe8jqqCOgZ3bxvtfGSOXgYoJ4DV6KOUWAzB1UqFX10eQ+hccO4cLAmIs+fceFoJUSJfq+BcDTGorYXJnC4GAL7jAFg+AAlgTiSkiBcBuC9AULGYAJSsFgzIrgLUVSaldx8QUhSAAZIQ4wFIeUCXI+JT87EZOse2ahazHFP2auBBOat3EsPTp4pgYz8kF0DiA-pISl7CJgaIupc9okoNiUs9p1VEl72SSPFZL8LHvwcRQ0pTTgAVKgKM2BEyTaNOaa865HSunqF6fC6qgzxjDMqQmEiYzalQv9jMsiLTAptPtlADyaSdGWL0Q-DZ6SqWZM0CJDkuzOQ2kThrAALCc3W-9Hm+KKbIlFRwnZCqgGpJ0-RMUFmqZCyJkyBXFO+QYkVSqmUst-qc3leLCmXN4aK2u0cBwNxEuK4UUrsWyp7vy3VWCSUGINdQI1lD1X7NcQwDlRyM5kq8eMuVFyB56tVS1FVhDy4OvruXU1krwVYuTDiykfL5U2tXnajyIb+EeXDU6tVDDm5JwAKzcvCXkxN-r-FXKDdVdNwTApRuJOauNlr4FJoDbat5PzV7VuWdVOtYKIVYKOOcoBya22isdpXStWDe0NplQOkt2q-FEsFZOgdXb3lTsQGamN0r40RKtS28tCzl2hqPeOpEY7Uksp2eSq9nJ1U5K1X64drbj0Zpalm5skbN3Rv7avIdOqX2KpPVgj9sdV7Tu3Raud-7F0GMWReiOis65GqPRB390HS3PsPXO+DK6QNIcNc2I9zKb05uyT63FT6APYdfTWsCoS7nFso-uk2zzpFwrw2BT5F71EbFMZlVZ9LrFMsEwC6lWSe3fvrZBxtDyF0Kto92wKDHgAiIHTBtjS7cPAYHdxzjgVeMFMIaJ2h+jyUmcBeoAxaHY2zvU5h6jWm14Ifurc1T9ycMabc057Tb6cN6Z03Owz-GSAWfE4yudFL-mmaBVgkj1646upbgwfNnqTlHDOQ52DgbAtgTXR2wKoHjUQLc2pvdzbWPeY47lwKAW-MtWC+fMLDLKHNesV+iV0n0O+pY1hpdqasH5ftQRx1n6PIqbK3+rLmm4POf01xneXyavVUa98trZm4vrdi+BqTfbbO7vnVR7LOG5vLeDROs71UiuzaOBNjzJ2vPIPY6e1p7absLaUUt+rOHVvGb+fY8L72osA5ay9mzO6XvMYq3197xK3uQ7yxd77J3rsI6OODqDJ2evQ8c7DxT66sdDbR3eoHAVV7MsbmTpuLjktwAAGxFsy-Jkdr3RV3Y8hQNoB3B3Taqy9gb-nFsubUSfPjTX-tbO25tiXQmv6zA5+0bnCaF0zf5-Dk7QgcHGMRCFohdKxOg8i1t9QPVc1MIYHTtLOAH3lfqcd-HBX6Olfu9ju3qvR3zbq3RgzoujNaJlwb9rHlNdGJMeLzZsupc-2NxFn+eaNYAHZGePt645rwOXkdit2zO+NMGFPKbc9MI9PKsfK6O+7rHXBibC49gOkvWPHtSJ82BLgZMa8arjve83cAE9W5t677hgd09Aczxj2TU3meAZuU9l3tunl87nVXlzreL1NwS3HmnScAAcGJs556H48FTllAr17n1EhfKb1eryX5xjvmq5Pl4vyP73Le2+3+0BBbvW+red33+bYfrmM+xajeMSJoAuEc1eK6K+laTcEEm+GsAAnLvo8H-v4gAezifkxodqngvE-g7vapActnfqfgPufk9v1lfu+m-kQSaP9rQYHmZp-uygwAgT-nvllgKndsAbzuQe7OAYMIQd9tATpsDpLlZloHAd3sTAAAzIE86T5oGPC4DGBHzqBpy3Qzr0G55Zb2BjRzLVaZ5HBoDACcCirEFYHyFHYAAiQgGwNGz+SmLeCgCgnSuAYKNBRwxojCzBms0hbBKBHBB+RwyhqhCg6he20qWhTaduuh40FBF6xhpht+URD+OBNhhAdh8R82XAzhrh7hmeGqXhbKegD0dQuA9oA08scAisPojwRggswA1YcMCMzQl480ZY-wE494TQScxMjgPgvRnqbYAB9RcoTR8MiMc014HRFYy0CoScZM-RtOZMQx-RdRAsYxsMExrRjQ7Rt4lY8x7iSxScbiqxIxGxQsWxLRUxpY+xcxPRnKxxnKZx6xDR4x1xbR0xdx3RzQGs+aTxKWLxRwoxlxzRkxnxtxi0XRnovxFuAJ9OQJBgFxjRVx4JuxXxUJd4MJScCe8Jvet0rCwxrxmxYJOxDoGJnRWJVY2+8J3+BJ6cRJwJyJ7xaJ5JkJlJBxDxLB8JrB9JwAjJSJbxqJZJV47JsxPxvR0h8JMhLxuIksv010xQo040k0MUJk0sH08eDAsY6AiEXAcAScqpsUwgu+Ze+60wAAREnBaRoIMLqRPAaXAF9mYkVCFNnBaSpNFMaaZHSBaVqFaRrDqXqQaRrEaSZDaRIUlknEGQ6YaVNNpHyXIWac2padaZbFwPafqYUE6SUpIT4bGKafkqmRrDacwNMKwBwNwLwK6Z3B6UZGqXFAKH6cWdqZmSGQwGGXFBGd4Qctqf4ZYeaVAAGQwBGUcLQOWWwJwDwB-lqXAPPPqQoHGQ2SaTgFzgERMi2XAKOadn5kTjWe6Z6fGfFLpM2aqLTjGVmaGUeaOIGW2Uud6ayhIfMG6UOYecuT6UiKecOXOQufeeGY+d3vOXAIuVee+f2cmSqt+duQLkThHHeTmd8hIc3EYIiGAJwIvEgP1BsMIAAIocxwDDA4DsEblDl4UMAgDblUAawMj0Fc4QUTpQXBQWGvn1nekFSnnkhJxAUgUdnXm07cV-lxS2lcW-mgXelv4iXAWCUmlr6pLy6YFXwf5JYoWYXoUsCYWny4X4UgDgVFmkXaWUXUXH5gQkFQ6QXWnB6UFVpI4v5wXBnZnOmZRMr0H64xbiHU404FlEXrlUYWmxiQjzwemR7iG+bBJ1lekmTsUu6QjTCQjuzzzTDzyWyyVgAGq5L-p+UBVBUMFR78EbD4DqCz70XiKZWBViGx5w6irhVHlRXMUxVQBxUeQJVQBJVMqNzuzVVgUnlFX1WNUtTNWtWJaPCyXqCKwRGY4Dk46lUwCRzZVuUVVQCpUEZM6+X+VlXBULWhWOGdVsXdV1WxUGIDWULtXDW5DU5fCQhJkZVrVzWWax4C60DGDIZjZjkTmVnTmPA7WRV7W5KQiQF-XuyQhv5A3JVKXNx-VXUOalW3WA6PAPWCHBK0DEzUEZrnWwRA2Q3arQ3lWWwPUo2I3I25nFHQDzyY2rVZU41w1X6PXPVgZPXMACBCDACWxfVxS1W5LzyQGc3xVv7zxkyg2RnNyc1k2p7Y0bW41X7GAI3xJS342bxo16B80i1TU3WU1bUpLGBy0a2E2qoK3QCxi6VQ2q3i1U3XI02EZgbjnMCTlVmWxI1m1a0dpI2O2Wz038CCDCCu0606ZS3XKa3XJLVRzpVG0U0m2VUrrm2jZgavXW3vWEDuz20R0u0mjO0opu2M2e0miy1p3S3vL+0rr5XdJFXXWh05UhX4EiSR0RqV1vVTnx2V252O5jnJ0tSp2cbp0e3M2jze2Z6+3t0t0-wC1HCjUaEyZ2aTX1LTWzVq1gSB3F0h3rVl2bU7kv5V3Zqt21220N2iq0AD2BRt3LYd1M0CI90v592H172PngRnXNylG3QVEGkjR6GdkrmiTEmgnbFKnP1HmwlwAXntlwAv38kDG3n2ViXfWpwMlrFMlCmklf0qk-2SU8WANHnwkCXgNs18kCkgkolwNP0IPvkOSoVqUaXYXABkUEVLEqVoUQAYVYVxQUM6VYPQOCkkmf34MTQ-2xhLEG3MPnGwPsPKmcOEPRnwm8OQP8lrFylaBAA), so you will be able to see how they actually works.